EIPS/eip-3541.md

```diff
@@ -0,0 +1,50 @@
+---
+eip: 3541
+title: Reject new contracts starting with the 0xEF byte
+author: Alex Beregszaszi (@axic), Paweł Bylica (@chfast), Andrei Maiboroda (@gumb0)
+discussions-to: https://ethereum-magicians.org/t/evm-object-format-eof/5727/4
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-03-16
+---
+
+## Abstract
+
+Disallow new code starting with the `0xEF` byte to be deployed. Code already existing in the account trie starting with `0xEF` byte is not affected semantically by this change.
+
+## Motivation
+
+Contracts conforming to the EVM Object Format (EOF), described by [EIP-3540](./eip-3540.md), are going to be validated at deploy time. In order to guarantee that every EOF-formatted contract in the state is valid, we need to prevent already deployed (and not validated) contracts from being recognized as such format. This will be achieved by choosing a byte sequence for the *magic* that doesn't exist in any of the already deployed contracts. To prevent the growth of the search space and to limit the analysis to the contracts existing before this fork, we disallow the starting byte of the format (the first byte of the magic).
```

Comment:
Because of this link need to get #3540 merged first.

Reply:
Alternatively, you could potentially reword this so it isn't dependent on 3540 specifically.  Maybe something like "having a reserved first byte will allow us to introduce EVM versioning in a future EIP."

Reply:
The EIP is not dependent on 3540 specification wise, but only in the motivation section. I think this is extremely important for understanding the motivation, so would prefer to keep it this way.

If/once 3540 goes live, we expect to simplify 3541.

Reply:
Having a reference to 3540 in the motivation will mean that this EIP cannot advance through the EIP process faster than EIP-3541.  I strongly recommend removing the reference so this EIP doesn't get blocked by that.  This EIP is definitely reasonable to stand on its own IMO, as having a mechanism for contract versioning is important no matter what future versioning EIP we go with.

Reply:
>  I strongly recommend removing the reference so this EIP doesn't get blocked by that. This EIP is definitely reasonable to stand on its own IMO, as having a mechanism for contract versioning is important no matter what future versioning EIP we go with.

Is that the current blocker to this eip getting merged?

Reply:
Historically I haven't blocked EIPs merging as draft for that, though I do strongly prefer people try to have their EIPs as high of quality as possible and as alignment with the repository patterns/practices before merging as draft.

The current blocker is that this PR is marked as a draft PR, which is why I haven't given it an official review yet.  😄

Reply:
@MicahZoltu What is the origin of the rule that every linked EIP is a dependency of the current one?
In EIP-1 I see only the requirement that all dependencies should be explicitly listed in `requires`, but no restrictions on what can be linked.

Reply:
> What is the origin of the rule that every linked EIP is a dependency of the current one?
> In EIP-1 I see only the requirement that all dependencies should be explicitly listed in `requires`, but no restrictions on what can be linked.

@gumb0 EIP authors linking to EIPs that never make it past draft and eventually get moved to withdrawn or stagnant.  Then 3 years later you have a final EIP that references completely irrelevant content.  EIPs should be written for *future* readers, not *current* core developers.

I think the target audience of EIPs is the major point of contention here.  Many viewthe EIP process as a change control process for Ethereum, rather than a standardization process.  If it were a change control process, then the utility of an EIP would end when it reached `Final`.  As a standardization process, the utility of an EIP is considered to *begin* when it reaches `Final`.  We have been talking a lot lately about revamping the Ethereum specification/standardization/change control process to help address this conflict, but at the moment EIPs are standards, not change control documents.

Reply:
I think it is exactly the opposite. The EIP **draft** is your best chance to properly explain and motivate a proposed EIP. This is because all other channels (Discord chat and ACD calls) are temporary and have very limited capacity. 
- An EIP draft should be written for _current_ core developers.
- Withdrawn or stagnant EIPs are relevant.

Reply:
The purpose of the `discussions-to` link in an EIP is to fulfill that need.  The EIP should be written as it would be when final, purely as a standard/specification.  The information included in the motivation/rationale section should be things that will be relevant to people in the future.

Reply:
The `discusssions-to` absolutely does not work for this. Nobody will ready 20+ posts there. Unless you are suggesting everyone should ask "What is the purpose of this EIP?" all over again. Moreover, the `FAQ` section is not allowed in an EIP. So we are gradually collecting answers for common questions and put then in Rationale. 

We have clear disagreement here. How do you plan to resolve it?

Reply:
IMO: the discussions-to is a place to discuss and nitpick over details in the eip. However, I do think that the EIP itself should contain a bit of history for "why do we want this eip". Even for future developers, it's highly relevant to understand the context in which an eip was created. and what preceded it.
The discussions-to can be used to discuss "hey that reference is not useful", or "can't we change param X to `1337` instead of `5`". When all is said and done, however, the EIP should stand alone, and any _useful_ parts of previous discussions may well be _inside_ the eip (in some abbreviated form)